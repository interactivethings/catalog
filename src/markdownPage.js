import {createElement, isValidElement} from 'react';
import Page from './components/Page/Page';

// This function simply intersperses the values between the strings, and
// passes the result as children to the Page component. No further
// transformation is done, the Page component itself processes strings
// as markdown.
//
// The most we could do at this point is to parse strings into MDAST (or
// a similar abstract form). We can't convert the markdown text into React
// elements because that requires the Catalog context.
//
// Values SHOULD be React Elements, strings, numbers, or anything
// stringifiable. The primary use case is to allow
// developers to easily instantiate React components in plain JavaScript,
// so that type checkers (flow, typescript) can verify that the correct
// props are pased to the component.
//
// > import {HintSpecimen, markdown} from 'catalog';
// > export const catalogPage = markdown`
// > # This is a page
// >
// > With a paragraph. And a number ${123}
// >
// > ${<HintSpecimen>And a hint</HintSpecimen>}
// >
// > ${<MyComponent isCustomComponent={'AWESOME'} />}
// > `;

const replaceLast = (f, arr) => arr.slice(0, -1).concat(f(arr[arr.length - 1]));

const markdownPage = (strings, ...values) => () =>
  createElement(Page, {},
    ...values.reduce((a, v, i) => {
      // If it's a valid React element, just concat to the end of the array
      if (isValidElement(v)) {
        return a.concat(v, strings[i + 1]);
      }

      // String-concat v to last and next string part
      if (typeof v === 'string' || typeof v === 'number') {
        return replaceLast(last => last + v + strings[i + 1], a);
      }

      // Finally, try to stringify v
      return replaceLast(last => last + JSON.stringify(v) + strings[i + 1], a);
    }, [strings[0]]
    ));

export default markdownPage;
