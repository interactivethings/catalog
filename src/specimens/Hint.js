import React, { PropTypes } from 'react';
import {Style} from 'radium';
import renderMarkdown from '../utils/renderMarkdown';
import Specimen from '../components/Specimen/Specimen';
import {text} from '../styles/typography';

function getStyle(theme) {
  return {
    container: {
      flexBasis: '100%'
    },
    hint: {
      ...text(theme),
      background: '#fff6dd',
      border: '1px solid #ffefaa',
      borderRadius: '2px',
      color: '#ffb400',
      padding: '20px'
    },
    neutral: {
      background: '#f9f9f9',
      color: '#666666',
      border: '1px solid #eee'
    },
    warning: {
      background: '#fff5f5',
      border: '1px solid #ffdddd',
      color: '#ee4040'
    },
    directive: {
      background: '#eafaea',
      border: '1px solid #bbebc8',
      color: '#2fbf62'
    }
  };
}

class Hint extends React.Component {
  render() {
    const {theme, children, warning, neutral, directive} = this.props;
    const styles = getStyle(theme);

    const warningStyle = warning ? styles.warning : null;
    const directiveStyle = directive ? styles.directive : null;
    const neutralStyle = neutral ? styles.neutral : null;

    return (
      <div style={styles.container}>
        <section style={{...styles.hint, ...warningStyle, ...directiveStyle, ...neutralStyle}} className='cg-Hint'>
          <Style
            scopeSelector='.cg-Hint'
            rules={{
              code: {
                display: 'inline',
                borderRadius: '2px',
                background: 'rgba(0,0,0,.03)',
                fontFamily: theme.fontMono,
                padding: '4px 5px',
                whiteSpace: 'pre-wrap'
              },
              ':first-child': {
                marginTop: 0
              },
              ':last-child': {
                marginBottom: 0
              }
            }}/>
            <div>{renderMarkdown({text: children})}</div>
          </section>
      </div>
      );
  }
}

Hint.propTypes = {
  children: PropTypes.string.isRequired,
  theme: PropTypes.object.isRequired,
  warning: PropTypes.bool,
  neutral: PropTypes.bool,
  directive: PropTypes.bool
};

export default Specimen(undefined, undefined, {withChildren: true})(Hint);
