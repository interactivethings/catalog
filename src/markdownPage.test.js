import React from "react";
import markdownPage from "./markdownPage";
import Hint from "./specimens/Hint";

test("Catalog markdown page template literal", () => {
  const page = markdownPage`
# This is a title

A paragraph

${(
    <Hint>
      {`# THIS IS A HINT

Cool, eh?`}
    </Hint>
  )}
`;
  expect(page).toMatchSnapshot();
});

test("Catalog markdown page template literal with arbitrary values", () => {
  const page = markdownPage`
# This is a title

A paragraph ${123}

~~~
${["foo", "bar", <div>hello</div>]}
~~~
`;
  expect(page).toMatchSnapshot();
});
