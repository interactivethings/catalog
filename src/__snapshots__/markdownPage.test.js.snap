// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog markdown page template literal 1`] = `
<Page>
  
# This is a title

A paragraph


  <SpecimenContainer>
    # THIS IS A HINT

Cool, eh?
  </SpecimenContainer>
  

</Page>
`;

exports[`Catalog markdown page template literal with arbitrary values 1`] = `
<Page>
  
# This is a title

A paragraph 123

~~~
[{"a":"some"},{"a":"data"}]
~~~

</Page>
`;
